tests/proxy/map_paging_predicate_test.py

```diff
@@ -0,0 +1,241 @@
+import os
+import unittest
+
+from tests.base import HazelcastTestCase
+from tests.util import configure_logging, get_abs_path, random_string
+from tests.custom_comparator import CustomComparator
+from hazelcast.serialization.predicate import PagingPredicate, is_greater_than_or_equal_to, is_less_than_or_equal_to, \
+    is_ilike
+from hazelcast import HazelcastClient
+from hazelcast.six import assertCountEqual
+from hazelcast.util import ITERATION_TYPE
+
+
+class MapPagingPredicateTest(HazelcastTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        configure_logging()
+
+    def setUp(self):
+        self.rc = self.create_rc()
+        self.cluster = self.create_cluster(self.rc, self._configure_cluster())
+        self.member1 = self.cluster.start_member()
+        self.member2 = self.cluster.start_member()
+        self.client = HazelcastClient()
+        self.map = self.client.get_map(random_string())
```

Comment:
I agree that making the map .blocking() does indeed solve unexpected failures. It also enables us to test without using assertTrueEventually() and without worrying about timeout. I initially tested it that way.

However, I changed it to the current version when I introduced synchronous paging support to be able to test it thoroughly. From my understanding, Python map methods should always return a future and should be able to function synchronously.

blocking() tests always pass so I could re-introduce them, however, if the current way fails sometimes as you mentioned, I am not sure if it's the desired behavior. Shouldn't map be able to handle two concurrent .values() calls? What would you recommend?

Reply:
Map infact could handle concurrent requests. But the problems are how the tests are structured. For example, in `test_get_3rd_then_next_page`, you do
```python
        self.map.values(paging) # async
        paging.next_page()
```
So, you make the API call with the page value of 3 but change the page afterward without waiting for the result of `.values()`.
However, when the server sends a response for the API call, on the `get_sorted_query_result_set`  method, you access the page again. It should have been 3 but you changed it to 4. That's a problem for the first call. It may return `46, 47` or an empty list. (You don't check that but I want to say it anyway)

Now, what fails the assertion is something similar. Assume a timing like this.

-> first call is sent (empty anchor list is sent to the server)
-> second call is sent (empty anchor list is sent to server since the response of the first call is not received yet)
-> result of the first call is received (server sends the first 4 pages, you set the anchor list accordingly, now the nearest page is 2)
-> result of second call is received (server sends the first 5 pages since you sent an empty anchor list. the nearest page is 2, according to the calculations you will output the elements at 2:4 as the output which is `42, 43` but it should have been `48, 49`. (Indexes 2:4 would be correct if the second call sends something like `[(0, (41, 41)), (1, (43, 43)), (2, (45, 45))]` as the anchor list to the server) That fails the assertion and the later assertions since the predicate is in an invalid state now)

So, the problem is sharing the predicate without any synchronization for the two concurrent calls

Reply:
Thanks a lot for the thorough explanation! That makes sense. Is this a problem that should somehow be fixed to accommodate for sharing the predicate for two concurrent calls or is it enough to change the map to blocking and disregard this case? I am not sure what the approach should be.

Reply:
I have changed the map to .blocking() in the test file for now. If you think that we can disregard the above case and testing with .blocking() is sufficient, I will also get rid of all the assertTrueEventuallys.

Reply:
It should be okay to change tests with `.blocking` maps and get rid of `assertTrueEventually`s. 

It is user's responsibility to make sure that they are using the necessary synchronization or logic for the shared data.